### DISCUSSION TITLE
Simple docker deployment following docker best practices

### DISCUSSION THREAD

**Discussion by banesullivan** [Kitware/trame#142] [Q&A] [answered]
The current method to bundle/deploy a trame app is too involved for a simple app. Let's say I have a simple trame app as a single Python file. How would I go about building that into a docker image via a single and standard `docker build` process?

The steps of building a virtual environment outside of a container then bundling that into a new image are counter to many best practices with Docker. I understand the motivation for this process, but it is too involved for most users and most examples. How can I bundle/deploy a trame app in a more standard process with a single Dockerfile and a single `docker build` command?

I assume this would be something like:

```Dockerfile
FROM kitware/trame

COPY myapp.py /deploy/myapp.py

... # Some trame command to run that script?
```

**Comment by banesullivan** [comment]
Ideally, I would like my base image to be `python:3.9-slim` rather than `kitware/trame` so that I have tighter control over the image size and Python version since the `kitware/trame` image does not (as far as I can tell) have tags for the python version (and using a non-tagged image like that is a bad practice)

**Reply by jourdain** [reply]
Then all the setup for multi-client will be non-existent.

**Comment by banesullivan** [comment]
Could I do something like the following?:

```Dockerfile
FROM python:3.9-slim
COPY myapp.py /deploy/myapp.py
RUN pip install trame
ENTRYPOINT ["python", "/deploy/myapp.py", "--server"]
```

or do I need to do some fancy things to build the launcher and and `www` directory?

**Reply by jourdain** [reply]
No because you bypass the launcher and multi-user setup.

**Comment by jourdain** [comment, marked as answer]
You can follow [that example](https://github.com/Kitware/trame/tree/master/examples/deploy/docker/SingleFile).

The code expect your app to be named `app.py` and any dependency to be describe inside the `setup/requirements.txt`.
